### PR TITLE
Backport.System.Threading.Lock implementation.

### PR DESCRIPTION
**Pull request by MarkCiliaVincenti**

### CODE REVIEW

**Comment by MarkCiliaVincenti** [comment]
For some reason I couldn't build this project on my machine, not even the main branch; loads of analyzer errors, so please ensure the changes are applied well.

**Comment by MarkCiliaVincenti** [comment]
I also changed LangVersion from preview to latest. You may or may want to revert this; if you added it for the System.Threading.Lock functionality, it's not needed anymore since .NET 9.0 was released.

**Comment by thomhurst** [comment]
Doesn't compile unfortunately:

`error CS0052: Inconsistent accessibility: field type 'Lock' is less accessible than field 'TestContext.Lock' [/home/runner/work/TUnit/TUnit/TUnit.Core/TUnit.Core.csproj::TargetFramework=net8.0]`

**Comment by MarkCiliaVincenti** [comment]
Try now please @thomhurst?

**Comment by thomhurst** [comment]
We're now changing the API and making the lock non-public

**Comment by MarkCiliaVincenti** [comment]
OK in this case I'll revert this particular lock then.

**Comment by MarkCiliaVincenti** [comment]
Reverted that part @thomhurst

**Comment by MarkCiliaVincenti** [comment]
I reverted back the LangVersion to preview.

**Comment by MarkCiliaVincenti** [comment]
Oops, missed a few LangVersion reverts.

**Comment by thomhurst** [comment]
Looks like you're actually changing most `latest` lang versions to `preview` now

**Comment by MarkCiliaVincenti** [comment]
> Looks like you're actually changing most `latest` lang versions to `preview` now

Yes I missed a few. You could move them to latest but you'd need to fix the pipelines so they compile with the latest SDK of .NET 9.0

**Comment by MarkCiliaVincenti** [comment]
All OK now?

File: TUnit.Analyzers/TUnit.Analyzers.csproj [hunk 1]
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <TargetFrameworks>netstandard2.0</TargetFrameworks>
     <Nullable>enable</Nullable>
-    <LangVersion>latest</LangVersion>
+    <LangVersion>preview</LangVersion>
```
Comment (thomhurst) [inline TUnit.Analyzers/TUnit.Analyzers.csproj]: @MarkCiliaVincenti here for example - Why does this need to be preview now?

Comment (MarkCiliaVincenti) [reply TUnit.Analyzers/TUnit.Analyzers.csproj]: Fixed.

**Comment by MarkCiliaVincenti** [comment]
@thomhurst not quite sure what this error relates to now, but likely unrelated to this PR.

**Comment by thomhurst** [comment]
Thanks @MarkCiliaVincenti !

**Pull request merged**